### ISSUE TITLE
VictoriaLogs: Crashing with `panic: runtime error: index out of range [0] with length 0` in `getBloomValuesWriterForColumnName`

### ISSUE THREAD

**Issue by Fusl** [VictoriaMetrics/VictoriaMetrics#7391]
### Describe the bug

VictoriaLogs randomly crashes with the following panic. For some reason, I can't reproduce it when feeding logs at a fast rate but the reproduction steps below feeding logs at a 1-second interval reliably reproduce the issue for me.

```
panic: runtime error: index out of range [0] with length 0

goroutine 205840 [running]:
github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage.(*streamWriters).getBloomValuesWriterForColumnName(0xc415d8?, {0xc000ee404d?, 0xc00573ea00?})
        github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage/block_stream_writer.go:165 +0xc6
github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage.(*column).mustWriteTo(0xc0034fc870, 0xc0045acd68, 0x4?)
        github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage/block.go:158 +0xea
github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage.(*block).mustWriteTo(0xc0040040f0, 0xc0040fc858, 0xc0042ae0e0, 0xc002542000, 0xc0025421c0)
        github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage/block.go:452 +0x206
github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage.(*blockStreamWriter).mustWriteBlockInternal(0xc002542000, 0xc0040fc858, 0xc0040040f0, 0x0)
        github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage/block_stream_writer.go:377 +0x21f
github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage.(*blockStreamWriter).MustWriteBlock(...)
        github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage/block_stream_writer.go:359
github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage.(*blockStreamWriter).MustWriteRows(0xc002542000, 0xc0040fc858, {0xc003354000, 0x4, 0x150}, {0xc000926008, 0x4, 0x155})
        github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage/block_stream_writer.go:336 +0xd8
github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage.(*blockStreamMerger).mustFlushRows(0xc0040fc820)
        github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage/block_stream_merger.go:264 +0x7e
github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage.mustMergeBlockStreams(0xc022551630, 0xc002542000, {0xc000637ab0, 0x2, 0x2}, 0x0)
        github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage/block_stream_merger.go:30 +0x129
github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage.(*datadb).mustMergeParts(0xc0001e0f00, {0xc000637a90, 0x2, 0x2}, 0x1)
        github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage/datadb.go:560 +0x656
github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage.(*datadb).mustMergePartsToFiles.func1({0xc000637a90?, 0xc0005ff7b0?, 0xc0001e0f00?})
        github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage/datadb.go:315 +0x5d
created by github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage.(*datadb).mustMergePartsToFiles in goroutine 3284
        github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage/datadb.go:309 +0x99
```

### To Reproduce

Window A (run `go run .` inside `app/victoria-logs`):
```
~/Projects/VictoriaMetrics/VictoriaMetrics/app/victoria-logs master ❯ go run .

2024-10-30T06:07:02.081Z        info    VictoriaMetrics/lib/logger/flag.go:12   build version: 
2024-10-30T06:07:02.081Z        info    VictoriaMetrics/lib/logger/flag.go:13   command-line flags
2024-10-30T06:07:02.081Z        info    VictoriaMetrics/app/victoria-logs/main.go:42    starting VictoriaLogs at "[:9428]"...
2024-10-30T06:07:02.081Z        info    VictoriaMetrics/app/vlstorage/main.go:64        opening storage at -storageDataPath=victoria-logs-data
2024-10-30T06:07:02.082Z        info    VictoriaMetrics/app/vlstorage/main.go:70        successfully opened storage in 0.001 seconds; smallParts: 0; bigParts: 0; smallPartBlocks: 0; bigPartBlocks: 0; smallPartRows: 0; bigPartRows: 0; smallPartSize: 0 bytes; bigPartSize: 0 bytes
2024-10-30T06:07:02.082Z        info    VictoriaMetrics/app/victoria-logs/main.go:50    started VictoriaLogs in 0.001 seconds; see https://docs.victoriametrics.com/victorialogs/
2024-10-30T06:07:02.082Z        info    VictoriaMetrics/lib/httpserver/httpserver.go:121        starting server at http://127.0.0.1:9428/
2024-10-30T06:07:02.082Z        info    VictoriaMetrics/lib/httpserver/httpserver.go:122        pprof handlers are exposed at http://127.0.0.1:9428/debug/pprof/
```

Window B (run `while true; do echo '{"foo":"'$RANDOM'","_msg":"x"}' | tee /dev/stderr | curl -fs -XPOST -T /dev/stdin "http://127.0.0.1:9428/insert/jsonline" || break; sleep 1; done`):
```
~ ❯ while true; do echo '{"foo":"'$RANDOM'","_msg":"x"}' | tee /dev/stderr | curl -fs -XPOST -T /dev/stdin "http://127.0.0.1:9428/insert/jsonline" || break; sleep 1; done

{"foo":"4045","_msg":"x"}
{"foo":"24624","_msg":"x"}
{"foo":"5943","_msg":"x"}
{"foo":"2266","_msg":"x"}
{"foo":"22879","_msg":"x"}
{"foo":"3254","_msg":"x"}
{"foo":"6699","_msg":"x"}
{"foo":"16830","_msg":"x"}
{"foo":"5847","_msg":"x"}
{"foo":"21261","_msg":"x"}
{"foo":"8409","_msg":"x"}
{"foo":"2221","_msg":"x"}
[... about 30 seconds of logs ...]
{"foo":"25031","_msg":"x"}
{"foo":"27025","_msg":"x"}
{"foo":"11220","_msg":"x"}
{"foo":"279","_msg":"x"}
{"foo":"3186","_msg":"x"}
{"foo":"22871","_msg":"x"}
~ ❯ 
```

Window A now shows:
```
2024-10-30T06:07:04.718Z        info    VictoriaMetrics/lib/memory/memory.go:42 limiting caches to 40271715532 bytes, leaving 26847810356 bytes to the OS according to -memory.allowedPercent=60
panic: runtime error: index out of range [0] with length 0

goroutine 260 [running]:
github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage.(*streamWriters).getBloomValuesWriterForColumnName(0xc5c438?, {0xc00013c088?, 0xc000126300?})
        /home/fusl/Projects/VictoriaMetrics/VictoriaMetrics/lib/logstorage/block_stream_writer.go:165 +0xc6
github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage.(*column).mustWriteTo(0xc0001f4f28, 0xc0002e2060, 0x2?)
        /home/fusl/Projects/VictoriaMetrics/VictoriaMetrics/lib/logstorage/block.go:158 +0xea
github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage.(*block).mustWriteTo(0xc0002b6050, 0xc0000f8858, 0xc000038310, 0xc00037e1e0, 0xc00037e3a0)
        /home/fusl/Projects/VictoriaMetrics/VictoriaMetrics/lib/logstorage/block.go:452 +0x206
github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage.(*blockStreamWriter).mustWriteBlockInternal(0xc00037e1e0, 0xc0000f8858, 0xc0002b6050, 0x0)
        /home/fusl/Projects/VictoriaMetrics/VictoriaMetrics/lib/logstorage/block_stream_writer.go:377 +0x21f
github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage.(*blockStreamWriter).MustWriteBlock(...)
        /home/fusl/Projects/VictoriaMetrics/VictoriaMetrics/lib/logstorage/block_stream_writer.go:359
github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage.(*blockStreamWriter).MustWriteRows(0xc00037e1e0, 0xc0000f8858, {0xc000130270, 0x2, 0x2}, {0xc00044e030, 0x2, 0x2})
        /home/fusl/Projects/VictoriaMetrics/VictoriaMetrics/lib/logstorage/block_stream_writer.go:336 +0xd8
github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage.(*blockStreamMerger).mustFlushRows(0xc0000f8820)
        /home/fusl/Projects/VictoriaMetrics/VictoriaMetrics/lib/logstorage/block_stream_merger.go:264 +0x7e
github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage.mustMergeBlockStreams(0xc000030230, 0xc00037e1e0, {0xc000442090, 0x2, 0x2}, 0x0)
        /home/fusl/Projects/VictoriaMetrics/VictoriaMetrics/lib/logstorage/block_stream_merger.go:30 +0x129
github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage.(*datadb).mustMergeParts(0xc0002e20c0, {0xc000442070, 0x2, 0x2}, 0x1)
        /home/fusl/Projects/VictoriaMetrics/VictoriaMetrics/lib/logstorage/datadb.go:560 +0x656
github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage.(*datadb).mustMergePartsToFiles.func1({0xc000442070?, 0xc0002557b0?, 0xc0002e20c0?})
        /home/fusl/Projects/VictoriaMetrics/VictoriaMetrics/lib/logstorage/datadb.go:315 +0x5d
created by github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage.(*datadb).mustMergePartsToFiles in goroutine 181
        /home/fusl/Projects/VictoriaMetrics/VictoriaMetrics/lib/logstorage/datadb.go:309 +0x99
exit status 2
```

### Version

victoria-logs-20241029-171200-tags-v0.38.0-victorialogs-0-g12223cf5d0

### Used command-line flags

```
-http.pathPrefix=/vlogs/
-retentionPeriod=8w
-enableTCP6=true
-search.maxQueryDuration=15m
-insert.maxLineSizeBytes=1048576
```

**Comment by Fusl** [comment]
As noted in Slack, swapping `ph.BloomValuesFieldsCount` with `ph.BloomValuesShardsCount` in 
https://github.com/VictoriaMetrics/VictoriaMetrics/blob/2ac07aa81316ad5eaa838eb9124f75f443798ee7/lib/logstorage/datadb.go#L539-L541
seems to fix this issue bug im not 100% sure if this is the right approach.

**Comment by p5i** [comment]
I dad the same issue after upgrading VictoriaLogs to the version v0.38.0.
The fix in this comment https://github.com/VictoriaMetrics/VictoriaMetrics/issues/7391#issuecomment-2446104623 helped me too, thanks @Fusl.

**Comment by valyala** [comment]
@Fusl , thanks for filing the bugreport! This panic triggers when VictoriaLogs merges parts, which contain logs only with constant fields. In this case `ph.BloomValuesFieldsCount` equals to 0 across all the parts (since it counts the number of non-const fields). This results in zero value for `bloomValuesShardsCount`, which is then passed to [`blockStreamWriter.MustInitForFilePart`](https://github.com/VictoriaMetrics/VictoriaMetrics/blob/258ee93fd17f4ba576c1beaa54eda0f01ec56759/lib/logstorage/block_stream_writer.go#L268). This, in turn, creates zero-length slice for `streamWriters.bloomValuesShards` at https://github.com/VictoriaMetrics/VictoriaMetrics/blob/258ee93fd17f4ba576c1beaa54eda0f01ec56759/lib/logstorage/block_stream_writer.go#L116 .

Later, when constant fields with different values are merged into the resulting field with these different values, the resulting field becomes non-constant. So it needs to be stored in the `streamWriters.bloomValuesShards` at https://github.com/VictoriaMetrics/VictoriaMetrics/blob/258ee93fd17f4ba576c1beaa54eda0f01ec56759/lib/logstorage/block_stream_writer.go#L165 . This results in the panic, since this slice is empty.

The fix is to make sure that the `streamWriters.bloomValuesShards` contains at least a single item. See the fix at 102e9d4f4ea7b421b082740c6f3ba0402eee1f84

**Comment by jiekun** [comment]
Wow, that was quick! I was planning to fix it after dinner :) Looks like there's no need anymore.

**Comment by Fusl** [comment]
@valyala Thank you for addressing this issue and @jiekun for the Docker image, I just rolled out the image on my end and will report back if I'm seeing any more issues.

**Comment by valyala** [comment]
The bugfix at https://github.com/VictoriaMetrics/VictoriaMetrics/commit/102e9d4f4ea7b421b082740c6f3ba0402eee1f84 has been included in [v0.39.0-victorialogs](https://docs.victoriametrics.com/victorialogs/changelog/). Closing the issue as fixed then.
